servers/src/main/java/tachyon/master/file/meta/InodeTree.java

```diff
@@ -276,48 +275,53 @@ public CreatePathResult createPath(TachyonURI path, CreatePathOptions options)
-      throw new FileAlreadyExistsException(path.toString());
-    }
-    if (options.isDirectory()) {
-      lastInode =
-          new InodeDirectory.Builder().setName(name)
-              .setId(mDirectoryIdGenerator.getNewDirectoryId())
-              .setParentId(currentInodeDirectory.getId()).setPersisted(options.isPersisted())
-              .build();
-      if (options.isPersisted()) {
-        String ufsPath = mMountTable.resolve(getPath(lastInode)).toString();
-        UnderFileSystem ufs = UnderFileSystem.get(ufsPath, MasterContext.getConf());
-        ufs.mkdirs(ufsPath, false);
+      } else {
+        LOG.info("FileAlreadyExistsException: " + path);
+        throw new FileAlreadyExistsException(path.toString());
       }
     } else {
-      lastInode =
-          new InodeFile.Builder().setBlockContainerId(mContainerIdGenerator.getNewContainerId())
-              .setBlockSizeBytes(options.getBlockSizeBytes()).setTTL(options.getTTL()).setName(name)
-              .setParentId(currentInodeDirectory.getId()).setPersisted(options.isPersisted())
-              .setCreationTimeMs(options.getOperationTimeMs()).build();
-      if (currentInodeDirectory.isPinned()) {
-        // Update set of pinned file ids.
-        mPinnedInodeFileIds.add(lastInode.getId());
+      if (options.isDirectory()) {
+        lastInode =
+            new InodeDirectory.Builder().setName(name)
+                .setId(mDirectoryIdGenerator.getNewDirectoryId())
+                .setParentId(currentInodeDirectory.getId()).build();
+        if (options.isPersisted()) {
+          toPersistInodes.add(lastInode);
+        }
+      } else {
+        lastInode =
+            new InodeFile.Builder().setBlockContainerId(mContainerIdGenerator.getNewContainerId())
+                .setBlockSizeBytes(options.getBlockSizeBytes()).setTTL(options.getTTL())
+                .setName(name).setParentId(currentInodeDirectory.getId())
+                .setPersisted(options.isPersisted()).setCreationTimeMs(options.getOperationTimeMs())
+                .build();
+        if (currentInodeDirectory.isPinned()) {
+          // Update set of pinned file ids.
+          mPinnedInodeFileIds.add(lastInode.getId());
+        }
       }
+      lastInode.setPinned(currentInodeDirectory.isPinned());
+
+      createdInodes.add(lastInode);
+      mInodes.add(lastInode);
+      currentInodeDirectory.addChild(lastInode);
+      currentInodeDirectory.setLastModificationTimeMs(options.getOperationTimeMs());
     }
-    lastInode.setPinned(currentInodeDirectory.isPinned());
 
-    createdInodes.add(lastInode);
-    mInodes.add(lastInode);
-    currentInodeDirectory.addChild(lastInode);
-    currentInodeDirectory.setLastModificationTimeMs(options.getOperationTimeMs());
+    for (Inode inode : toPersistInodes) {
+      String ufsPath = mMountTable.resolve(getPath(inode)).toString();
+      UnderFileSystem ufs = UnderFileSystem.get(ufsPath, MasterContext.getConf());
+      ufs.mkdirs(ufsPath, false);
```

Comment:
Is there a way to use mkdirs() with recursive flag, so that we only have to mkdirs() for the last (lowest) directory? It might be more efficient that way too.